light-client/src/components/verifier.rs

```diff
@@ -55,75 +56,184 @@ pub trait Verifier: Send + Sync {
+        // Ensure the header validator hashes match the given validators
+        verdict!(self.predicates.validator_sets_match(
+            untrusted.validators,
+            untrusted.signed_header.header.validators_hash,
+            &self.hasher,
+        ));
+
+        // TODO(thane): Is this check necessary for IBC?
+        if let Some(untrusted_next_validators) = untrusted.next_validators {
+            // Ensure the header next validator hashes match the given next validators
+            verdict!(self.predicates.next_validators_match(
+                untrusted_next_validators,
+                untrusted.signed_header.header.next_validators_hash,
+                &self.hasher,
+            ));
+        }
+
+        // Ensure the header matches the commit
+        verdict!(self.predicates.header_matches_commit(
+            &untrusted.signed_header.header,
+            untrusted.signed_header.commit.block_id.hash,
+            &self.hasher,
+        ));
+
+        // Additional implementation specific validation
+        verdict!(self.predicates.valid_commit(
+            untrusted.signed_header,
+            untrusted.validators,
+            &self.commit_validator,
+        ));
+
+        // Check that the untrusted block is more recent than the trusted state
+        verdict!(self
+            .predicates
+            .is_monotonic_bft_time(untrusted.signed_header.header.time, trusted.header_time,));
+
+        let trusted_next_height = trusted.height.increment();
+
+        if untrusted.height() == trusted_next_height {
+            // If the untrusted block is the very next block after the trusted block,
+            // check that their (next) validator sets hashes match.
+            verdict!(self.predicates.valid_next_validator_set(
+                untrusted.signed_header.header.validators_hash,
+                trusted.next_validators_hash,
+            ));
+        } else {
+            // Otherwise, ensure that the untrusted block has a greater height than
+            // the trusted block.
+            verdict!(self
+                .predicates
+                .is_monotonic_height(untrusted.signed_header.header.height, trusted.height));
+
+            // Check there is enough overlap between the validator sets of
+            // the trusted and untrusted blocks.
+            verdict!(self.predicates.has_sufficient_validators_overlap(
+                untrusted.signed_header,
+                trusted.next_validators,
+                &options.trust_threshold,
+                &self.voting_power_calculator,
+            ));
+        }
+
+        // Verify that more than 2/3 of the validators correctly committed the block.
+        verdict!(self.predicates.has_sufficient_signers_overlap(
+            untrusted.signed_header,
+            untrusted.validators,
+            &self.voting_power_calculator,
+        ));
+
+        Verdict::Success
     }
 }
+
+/// The default production implementation of the [`PredicateVerifier`].
```

Comment:
This implies that there is a common understanding what a "production" environment constitutes. More qualification would aid in contextualised this.

Reply:
Yes this is something we've been talking/thinking about for a year now :smile: We're definitely open to suggestions on the naming front.

The original purpose, afaik, was to disambiguate mock versions from non-mock versions of the types, but I think @romac can elaborate on this better than I can?

Reply:
Yeah I am not super happy with the current terminology (that I came up with). I guess we need to differentiate between:
a) the trait (previously `Predicates`, now `Verifier`)
b) the "production" implementation (previously `ProdPredicates`, now `ProdVerifier`)
c) the potential mock implementations of the trait (no use for that yet)

I agree that the `Prod` prefix is subpar, maybe `Default` or `Standard` would be a better fit?